RecoMuon/TrackerSeedGenerator/plugins/TSGForOIDNN.cc

```diff
@@ -0,0 +1,974 @@
+                                                                ed2 = mymeas_next.end();
+             imea_next != ed2;
+             ++imea_next) {
+          // save valid measurements
+          if (imea_next->recHit()->isValid())
+            meas_next.push_back(*imea_next);
+
+        }  // end loop over mymeas_next
+      }    // end loop over dets_next
+
+      // sort valid measurements found on this layer
+      std::sort(meas_next.begin(), meas_next.end(), TrajMeasLessEstim());
+
+      // loop over valid measurements compatible with updated TSOS (TSOS updated with a hit on the first layer)
+      for (std::vector<TrajectoryMeasurement>::const_iterator mea_next = meas_next.begin(); mea_next != meas_next.end();
+           ++mea_next) {
+        if (nmeas >= max_meas)
+          break;  // skip if we already found enough hits
+
+        // try to update TSOS again, with an additional hit
+        updatedTSOS_next = updator_->update(mea_next->forwardPredictedState(), *mea_next->recHit());
+
+        if (not updatedTSOS_next.isValid())
+          continue;  // skip if TSOS updated with additional hit is not valid
+
+        // If there was a compatible hit on this layer, we end up here.
+        // An additional compatible hit is saved.
+        seedHits.push_back(*mea_next->recHit()->hit());
+        det_id = mea_next->recHit()->geographicalId().rawId();
+        nmeas++;
+        found_compatible_on_next_layer++;
+
+      }  // end loop over meas_next
+
+      addtnl_layers_scanned++;
+
+    }  // end loop over compLayers (additional layers scanned after the original layer)
+
+    if (found_compatible_on_next_layer == 0)
+      continue;
+    // only consider the hit if there was a compatible hit on one of the additional scanned layers
+
+    // Create a seed from two saved hits
+    PTrajectoryStateOnDet const& pstate = trajectoryStateTransform::persistentState(updatedTSOS_next, det_id);
+    TrajectorySeed seed(pstate, std::move(seedHits), oppositeToMomentum);
+
+    LogTrace("TSGForOIDNN") << "TSGForOIDNN::makeSeedsFromHitDoublets: Number of seedHits: " << seedHits.size()
+                            << std::endl;
+    out.push_back(seed);
+
+    found++;
+    numSeedsMade++;
+    hitDoubletSeedsMade++;
+
+    if (found == numOfHitsToTry_)
+      break;  // break if enough measurements scanned
+    if (hitDoubletSeedsMade > maxHitDoubletSeeds_)
+      return;  // abort if enough seeds created
+
+  }  // end loop over measurements compatible with original TSOS
+
+  if (found)
+    layerCount++;
+}
+
+//
+// Prepare a dictionary of variables to use as input features for DNN
+//
+std::map<std::string, float> TSGForOIDNN::getFeatureMap(reco::TrackRef l2,
+                                                        const TrajectoryStateOnSurface& tsos_IP,
+                                                        const TrajectoryStateOnSurface& tsos_MuS) const {
+  std::map<std::string, float> the_map;
+  the_map["pt"] = l2->pt();
```

Comment:
IIUC, the list of features is fixed; it may be more practical to use `unordered_map` and make it once per event and reuse in the `::produce` calls and pass it by reference for filling or evaluation/reading.

Reply:
Done!